src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs

```diff
@@ -568,6 +500,39 @@ public void Dispose()
             _httpClientLock.Dispose();
         }
 
+        private class ThrottledResponse : IDisposable
+        {
+            private IThrottle _throttle;
+            private readonly object _throttleLock = new object();
+
+            public ThrottledResponse(IThrottle throttle, HttpResponseMessage response)
+            {
+                if (response == null)
+                {
+                    throw new ArgumentNullException(nameof(response));
+                }
+
+                _throttle = throttle;
+                Response = response;
+            }
+            
+            public HttpResponseMessage Response { get; }
+
+            public void Dispose()
+            {
+                Response.Dispose();
+
+                lock (_throttleLock)
+                {
+                    if (_throttle != null)
```

Comment:
Not sure if I understand your point. The wrapper object is so that we can conveniently dispose both the `HttpResponseMessage` the release the lock with a `using`. The `null` check is to ensure double disposes don't release twice.